RELEASES.md

```diff
@@ -20,6 +20,121 @@ v4.5.0 (development in progress)
   ```
   [PR #2821](https://github.com/leanprover/lean4/pull/2821) and [RFC #2838](https://github.com/leanprover/lean4/issues/2838).
 
+* The syntax of the `termination_by` and `decreasing_by` termination hints is overhauled:
+
+  * They are now placed directly after the function they apply to, instead of
+    after the whole `mutual` block.
+  * Therefore, the function name no longer has to be mentioned in the hint.
+  * If the function has a `where` clause, the `termination_by` and
+    `decreasing_by` for that function clauses come before the `where`. The
+    functions in the `where` clause can have their own termination hints.
+  * The `termination_by` clause must bind exactly those parameters that are not
+    already bound by the function header. If there are none, the `=>` can be
+    omitted.
```

Comment:
Does this mean that
```
def f (x : Nat) : Nat -> Nat :=
  if h : x < 4 then id else fun n => 1 + f (x - 10) n
termination_by x
```
is forbidden?

I asked this in the other comment but I think it got lost in the shuffle. If it's allowed, then perhaps the text could be:
```suggestion
  * The `termination_by` clause may bind parameters that are not
    already bound by the function header and are used in the termination
    argument. If there are none, the `=>` must be omitted.
```

Reply:
It 's forbidden, correct.


```suggestion
  * The `termination_by` clause must bind exactly those parameters that are not
    already bound by the function header. If there are none, the `=>` must be
    omitted.
```

Reply:
I think that example should not be forbidden, since it doesn't introduce the variable. Banning it would mean that specializing an opaque type can cause it to fail:
```lean
opaque A : Type := Nat → Nat
def f (x : Nat) (a : A) : A :=
  if h : x = 0 then a else f (x - 1) a
termination_by x -- works

def B : Type := Nat → Nat
def f (x : Nat) (a : B) : B :=
  if h : x = 0 then a else f (x - 1) a
termination_by x -- fails
```

Reply:
It’s not the type that matters, but rather the function definition (number of manifest lambdas after the `:=`, or scrutinees in the cases syntax. I should clarify that.

The rationale is that that the implementation has to figure out _some_ arity of the function, and arity affects, for example, how arguments are tupled. It seems better to require the user to explicitly acknowledge the arity as inferred and used by the system. It’s a little bit of extra red tape for the user, but it takes away a pointless choice (so less for the user to decide, which is good) and gives more insight in the internals.

I would support allowing `termination_by` with less paramters if that was _prescriptive_, i.e. really affected the argument packing as well, if there was a compelling use case for that, but that’d be a separate change.

Reply:
> gives more insight in the internals

I would call that "coupling to internal details" and I don't think it's a good thing. I don't see any reason for this either. If the user only introduces 2 arguments in the `termination_by` clause, then you should only consider those two arguments while reasoning about termination, you shouldn't go ahead and consider 3 arguments and then complain at the user for having done so.

Reply:
I agree that the implementation (existing already) leaks heuristic implementation detail, but fixing that isn't the point of this PR, and I don't want to inflate it more.

And given that the implementation has leaky heuristics, let's be honest about it and reserve “binding fewer parameters” as the syntax for a possible future where this _does_ guide/replace the heuristic.

You are right about ”manifest” being the wrong word.

Reply:
By the way, possibly related, I think it is weird that `y` is in scope in the second `termination_by` hint. It is outside the scope of the lambda, so I should not be allowed to write `termination_by _ => y` there, but the current implementation allows me to (only in the second case). So maybe that gives an alternative answer for your arity problem: only stuff left of the colon is in scope, stuff in the body must be introduced explicitly using the `=>` syntax, no matter whether there are manifest lambdas or match statements or anything like that.

Reply:
> The existing implementation does not require any particular number of arguments to be used. 

Well, it syntactically allows you to omit some. But this is very superficial, and does not affect heuristic about how many parameters are actually put in the packed argument in the wf transformation.

The two functions above are translated differently; the first into a recursion on `Nat` and the second into a recursion on `Nat × Nat`.

I agree that this heuristic is maybe suboptimal and this will become more relevant once we infer induction principles from function definitions, which will look quite different depending on the number of paramteres in the recursion. So giving the user some control over this is desirable!

A good syntax for “how many parameters should be packed” is, I’d say, “the parameters given in `termination_by`”. The implementation in this PR is nicely forward-compatible with that interpretation, by only allowing the syntax that matches the (currently implementation defined) heuristic.

The old behavior accepts `termination_by f' => x` and `termination_by f' x n => x`, but there is no actual difference in meaning between these two syntaxes, so to me it’s preferrable to allow only one. Less for the style guide writers to discuss.

Reply:
I have strong reservations about a system for which the interface is "guess what the compiler's internals do, and if you don't guess right, then too bad on you". I guess it's fine as a stopgap, but then we should acknowledge this.

In my opinion, if there's an internal notion of function arity that user's can't directly specify, then the interface shouldn't require users to figure out what it is and stick to it. This is fragile, in that changes to elaboration that affect function arity become breaking changes, and it's obscure, in that the core language elaboration leaks into the UI.

Why can't the interface be that users can specify as many of the further arguments as they need in order to specify the termination measure? I don't get why "the internals might be different" is an argument for the current state of things.

Reply:
(basically, I agree with @digama0's suggestion of "only stuff left of the colon is in scope, stuff in the body must be introduced explicitly using the => syntax, no matter whether there are manifest lambdas or match statements or anything like that")

Reply:
> The two functions above are translated differently; the first into a recursion on `Nat` and the second into a recursion on `Nat × Nat`.

By the way, I think this is suboptimal from a proof perspective. The second argument should not participate in the recursion at all, this is a definition by recursion on the first argument (a recursion on `Nat`) which has a motive of type `Nat -> Nat`. Some day I'll go figure out why it's generating a bad proof, if you don't beat me to it.

Reply:
> Why can't the interface be that users can specify as many of the further arguments as they need in order to specify the termination measure? I don't get why "the internals might be different" is an argument for the current state of things.

The current syntax on `master` is rather full of surprises. Consier:
```
def foo (n : Nat) : Nat → Nat → Nat
 | a, b => …
```
which parameter do you think is by each of these:
```
termination_by foo => n        -- 1
termination_by foo n => n      -- 2
termination_by foo _ n => n    -- 3
termination_by foo n _ => n    -- 4
termination_by foo _ _ n => n  -- 5
termination_by foo m => n + m  -- 6
```
All of them are valid! But this is very confusing, especially for the reader of the code.

The confusion is reduced if you _know_ that before the `=>` are exactly and only those obscure “extra parameters” inferred by the system. With the current PR, these 5 arguments are written as
```
termination_by _ _ => n
termination_by _ n => n
termination_by _ n => n
termination_by n _ => n
termination_by _ n => n
termination_by _ m => n + m
```

> “only stuff left of the colon is in scope”

That's the case. Old behavior was that you can rename stuff on the left of the colon (4), new behavior is that you cannot rename stuff on the left (because why should you besides to confuse the reader).

> “stuff in the body must be introduced explicitly using the => syntax”

agreed, and that’s the case.

> “no matter whether there are manifest lambdas or match statements or anything like that")”

That’s not possible at the moment: The system has its current rules for how many parameters there are, and this is not just an implementation detail. It certainly gives an _upper_ limit on the number of parameters here; it may show up in proofs; and will certainly show up once we derive induction principles from function definitions. So I don’t think it’s necessary to hide this detail.

As I said above: We can discuss changing this implementation, and I am sympathetic (up to prioritization of features), but maybe in a separate RFC. I’d say changing that is out of scope for this PR, but at least this PR’s syntax and semantics is a subset of that, so adding these features later will be a strict extension. That’s a good thing, isn’t it?

Reply:
It seems unnecessarily obfuscatory to be using the same variable `n` there when there is already an `n` in scope. In the "left of the colon" proposal, those examples would be expressible as
```lean
-- 1
termination_by n
termination_by _ => n
termination_by _ _ => n
-- 2, 3, 5
termination_by _ b => b
-- 4
termination_by a => a
termination_by _ a => a
-- 6
termination_by _ m => n + m
```

> > “no matter whether there are manifest lambdas or match statements or anything like that")”
> 
> That’s not possible at the moment: The system has its current rules for how many parameters there are, and this is not just an implementation detail. It certainly gives an _upper_ limit on the number of parameters here; it may show up in proofs; and will certainly show up once we derive induction principles from function definitions. So I don’t think it’s necessary to hide this detail.

It should ideally not show up in proofs. In fact, if I recall correctly that was a major motivation for removing `termination_by'`, it was said to be too coupled to the implementation because you could see the tupled type explicitly, and this is true. We should hide this detail, because I fully expect it to change when we get better about not adding unnecessary things to the tuple. Maybe we can't do this perfectly, but at least we should avoid the common case having to couple to the implementation detail, and requiring that all arguments in the lambdaTelescope be introduced causes exactly that coupling.

Reply:
Yes, I was sloppy in the example and omitted the `_`.


Ok, implemented in 4a5b39a442997d836e6854f9fc5fc3e577216ecb, and updated the RELEASE notes.